crypto/fastforward-0.51/fastforward.1
.TH fastforward 1
.SH NAME
fastforward \- forward mail according to a cdb database
.SH SYNOPSIS
in
.BR .qmail-default :
.B | fastforward
[
.B \-nNpPdD
]
.I cdb
.SH DESCRIPTION
.B fastforward
forwards each incoming message
according to instructions in
.I cdb
created by
.BR setforward .

If there is no forwarding instruction in
.I cdb
for the incoming recipient address,
.B fastforward
will bounce the message.

You can override
.B .qmail-default
with a specific
.BR .qmail-\fIrecipient ;
see
.BR dot-qmail (5).

Warning to system administrators:
Messages do not reach
.B ~alias/.qmail-default
unless they are controlled by the
.B alias
user.
See
.BR qmail-getpw (8).

.B SECURITY WARNING:
If
.I cdb
includes instructions pointing to a mailing list owned by another user,
that user gains some amount of control over
.BR fastforward 's
behavior.
In particular, he can force
.B fastforward
to open any file that you can access,
and to read any world-readable file that you own,
even if the file is in a world-inaccessible directory.
.SH "OPTIONS"
.TP 5
.B \-n
No delivery.
.B fastforward
will print a description of its actions,
but will not actually read or forward a message.
.TP
.B \-N
(Default.)
Forward a message as usual.
.TP
.B \-p
Pass through.
If
.B fastforward
does not find the recipient in
.IR cdb ,
it exits 0,
giving the message to further commands in
.BR .qmail-default .
If
.B fastforward
finds the recipient,
it forwards the message and exits 99,
so that further commands are skipped.
.TP
.B \-P
(Default.)
Do not pass through.
If
.B fastforward
finds the recipient,
it forwards the message and exits 0.
Otherwise it bounces the message.
.TP
.B \-d
Use
.B $DEFAULT@$HOST
as the recipient address, or
.B $EXT@$HOST
if
.B $DEFAULT
is not set.
.TP
.B \-D
(Default.)
Use
.B $RECIPIENT
as the recipient address.
.SH VERSION
This is
.B fastforward
0.51.
The
.B fastforward
home page is
.BR http://pobox.com/~djb/fastforward.html .
.SH "SEE ALSO"
newaliases(1),
printforward(1),
setforward(1),
dot-qmail(5),
qmail-command(8),
qmail-local(8),
qmail-getpw(8)


sc/CHANGES
CHANGES BETWEEN 7.14 and 7.15:
Chuck Martin
	-Fixed "``" and "''" bug, introduced in 7.14 (they weren't being
	 accepted).  I guess 7.14 was released too soon after last minute
	 changes, without sufficient testing.
	-Fixed format bug that caused segfaults when there was no decimal
	 point in the format string, also introduced in 7.14 last minute
	 changes.
	-When entering insert mode with '+' or '-', and the cell already
	 contains a numeric value or expression, the edit is now saved in
	 the dot buffer, as it should be.
	-Don't save the current position (for the "`" and "'" commands)
	 during a goto in a macro.  Also, don't save the current position
	 during any movement command where the current cell doesn't change
	 (leave the saved position pointing to the last position before an
	 actual move).

CHANGES BETWEEN 7.13 and 7.14:
Chuck Martin
	-Added a braille option ("set braille" or ^Tb) with two different
	 modes, which makes use of braille displays with sc much easier
	 by judicious placement of the hardware cursor.
	-The "cellcur" option (current cell highlighting) is no longer
	 saved with a file, since it's more likely a user preference
	 and/or terminal dependent, and is more appropriate for use in a
	 user's .scrc.
	-Split the last fprintf() in list_frames() so that the outside
	 and inside ranges will be different (as they should be), even if
	 neither is a named range, since r_name() uses a single static
	 buffer.
	-The range copy command (/c) with only one argument will now use the
	 highlighted range as the source if one is highlighted.  Otherwise,
	 it will copy the last copied range, as before.
	-The range copy command (/c) with no arguments will now use the
	 highlighted range as the destination, if one is highlighted, and
	 the last copied range as the source.
	-The addnote command with only one argument will now use the
	 highlighted range if there is one, and the current cell otherwise,
	 as the range to link to the cell specified by the single argument.
	-The frame command with a single argument will use the highlighted
	 range, if any, as the inside range, and the single argument as
	 the outside range.  If no range is highlighted, the single argument
	 will be used as the inner range, and the existing outer range
	 containing the current cell will be used as the outer range, which
	 which was the previous behavior whenever a single range was speci-
	 fied.  If the current cell is not in an existing framed range, an
	 error message will be issued.
	-The frame command may now be used without any arguments, provided
	 a range is highlighted (which will be used as the inner range)
	 and the current cell is in a framed range whose outer range is to
	 be used as the default outer range.
	-Fixed a bug that prevented defining a new range with a name that
	 was the same as the beginning of a previously defined longer
	 range name.
	-A new command line option, -a, will skip the autorun macro if
	 one is defined.
	-Bug fix: if a cell contains a string expression, but no numeric
	 value, pressing "e" to edit the non-existent numeric value will
	 no longer bring up the string expression for editing (it wouldn't
	 work, anyway, because string expressions don't work with the "let"
	 command).
	-You can now use the cursor keys with ^Tr to set craction.
	-In numeric mode, entering insert mode by pressing a digit, '+',
	 or '-' will cause the cursor keys to terminate an entry and
	 move the cell cursor in the desired direction.  Switching to
	 edit mode will change the cursor keys back to their normal
	 function (moving the cursor in the command line or history), even
	 after switching back to insert mode.
	-Fixed @index function so that indexing by both row and column
	 will still work even if there is only one row or one column
	 (so that, for example, a macro could generically access the
	 cell in the first row and first column of a range without
	 regard for the actual dimensions of the range).
	-All flags are turned off in the clearent() function except
	 is_changed and is_cleared, which are turned on.  Also, the
	 is_cleared flag is turned off in the fill() function after the
	 cell is cleared and then assigned a new value.  This prevents
	 errors from appearing in expressions which reference cells in
	 the filled range whenever the sync_refs() function is called,
	 which happens any time any cell or range of cells is yanked,
	 deleted, pulled, copied, or moved.
	-Each instance of @ext now saves its own previous value, which
	 eliminates the longstanding bug that caused the last returned
	 value to be used for all instances if external functions were
	 enabled and later disabled.
	-Added a % command in edit mode that finds the matching parenthesis
	 of a pair, as in vi (only works for parentheses, not for braces
	 or brackets).
	-When starting a new command, the dot buffer will not be altered
	 unless the command starts out in edit mode (like the "e" and
	 "E" commands).  All other commands will only alter the dot buffer
	 after entering edit mode for the first time.  This allows things
	 like editing data, using "goto" to jump to a new cell, and then
	 "." to make the same change to the new cell (previously, the
	 "goto" command caused the previous edit to be lost from the dot
	 buffer).
	-Numeric arguments are now saved for use with the dot buffer when
	 editing a line, so the same argument will be used again when
	 repeating the action with the "." command.
	-If a command line is identical to the last command in the history,
	 it won't be stored in the history again.
	-The copy and move commands (/c and /m) no longer have an effect
	 on the delete buffer.  This allows you to do things like delete
	 a cell or range of cells, move things around in the spreadsheet,
	 and then pull the deleted cell(s) back in.
	-Implemented named buffers, "a" through "z" and "0" through "9",
	 selected with the " key.  Buffers "1" through "9" hold the last

crypto/fastforward-0.51/fastforward.c
#include <sys/types.h>
#include <sys/stat.h>
#include "stralloc.h"
#include "substdio.h"
#include "subfd.h"
#include "strset.h"
#include "sgetopt.h"
#include "readwrite.h"
#include "exit.h"
#include "strerr.h"
#include "env.h"
#include "sig.h"
#include "qmail.h"
#include "fmt.h"
#include "case.h"
#include "alloc.h"
#include "coe.h"
#include "seek.h"
#include "wait.h"
#include "fork.h"

#define FATAL "fastforward: fatal: "

void usage()
{
  strerr_die1x(100,"fastforward: usage: fastforward [ -nNpP ] data.cdb");
}
void nomem()
{
  strerr_die2x(111,FATAL,"out of memory");
}

void print(s)
char *s;
{
  char ch;
  while (ch = *s++) {
    substdio_put(subfderr,&ch,1);
  }
}

void printsafe(s)
char *s;
{
  char ch;
  while (ch = *s++) {
    if (ch < 32) ch = '_';
    substdio_put(subfderr,&ch,1);
  }
}

struct qmail qq;
char qp[FMT_ULONG];
char qqbuf[1];

int qqwrite(fd,buf,len) int fd; char *buf; int len;
{
  qmail_put(&qq,buf,len);
  return len;
}

substdio ssqq = SUBSTDIO_FDBUF(qqwrite,-1,qqbuf,sizeof qqbuf);

char messbuf[4096];
substdio ssmess = SUBSTDIO_FDBUF(read,0,messbuf,sizeof messbuf);

int flagdeliver = 1;
int flagpassthrough = 0;

char *dtline;
stralloc sender = {0};
stralloc programs = {0};
stralloc forward = {0};

strset done;
stralloc todo = {0};

stralloc mailinglist = {0};

void dofile(fn)
char *fn;
{
  int fd;
  struct stat st;
  int i;
  int j;

  if (!stralloc_copys(&mailinglist,"")) nomem();

  fd = open_read(fn);
  if (fd == -1)
    strerr_die4sys(111,FATAL,"unable to read ",fn,": ");
  if (fstat(fd,&st) == -1)
    strerr_die4sys(111,FATAL,"unable to stat ",fn,": ");
  if ((st.st_mode & 0444) != 0444)
    strerr_die3x(111,FATAL,fn," is not world-readable");
  if (slurpclose(fd,&mailinglist,1024) == -1)
    strerr_die4sys(111,FATAL,"unable to read ",fn,": ");

  i = 0;
  for (j = 0;j < mailinglist.len;++j)
    if (!mailinglist.s[j]) {
      if ((mailinglist.s[i] == '.') || (mailinglist.s[i] == '/')) {
        if (!stralloc_cats(&todo,mailinglist.s + i)) nomem();
        if (!stralloc_0(&todo)) nomem();
      }
      else if ((mailinglist.s[i] == '&') && (j - i < 900)) {
        if (!stralloc_cats(&todo,mailinglist.s + i)) nomem();
        if (!stralloc_0(&todo)) nomem();
      }
      i = j + 1;
    }
}

char *fncdb;
int fdcdb;
stralloc key = {0};
uint32 dlen;
stralloc data = {0};

void cdbreaderror()
{
  strerr_die4sys(111,FATAL,"unable to read ",fncdb,": ");
}

int findtarget(flagwild,prepend,addr)
int flagwild;
char *prepend;
char *addr;
{
  int r;
  int at;

  if (!stralloc_copys(&key,prepend)) nomem();
  if (!stralloc_cats(&key,addr)) nomem();
  case_lowerb(key.s,key.len);

  r = cdb_seek(fdcdb,key.s,key.len,&dlen);
  if (r == -1) cdbreaderror();
  if (r) return 1;

  if (!flagwild) return 0;
  at = str_rchr(addr,'@');
  if (!addr[at]) return 0;

  if (!stralloc_copys(&key,prepend)) nomem();
  if (!stralloc_cats(&key,addr + at)) nomem();
  case_lowerb(key.s,key.len);

  r = cdb_seek(fdcdb,key.s,key.len,&dlen);
  if (r == -1) cdbreaderror();
  if (r) return 1;

  if (!stralloc_copys(&key,prepend)) nomem();
  if (!stralloc_catb(&key,addr,at + 1)) nomem();
  case_lowerb(key.s,key.len);

  r = cdb_seek(fdcdb,key.s,key.len,&dlen);
  if (r == -1) cdbreaderror();
  if (r) return 1;

  return 0;
}

int gettarget(flagwild,prepend,addr)
int flagwild;
char *prepend;
char *addr;
{
  if (!findtarget(flagwild,prepend,addr)) return 0;

  if (!stralloc_ready(&data,(unsigned int) dlen)) nomem();
  data.len = dlen;
  if (cdb_bread(fdcdb,data.s,data.len) == -1) cdbreaderror();

  return 1;
}

void doprogram(arg)
char *arg;
{
  char *args[5];
  int child;
  int wstat;

  if (!flagdeliver) {
    print("run ");
    printsafe(arg);
    print("\n");
    substdio_flush(subfderr);
    return;
  }

  if (*arg == '!') {
    args[0] = "preline";
    args[1] = "sh";
    args[2] = "-c";
    args[3] = arg + 1;
    args[4] = 0;
  }
  else {
    args[0] = "sh";
    args[1] = "-c";
    args[2] = arg + 1;
    args[3] = 0;
  }

  switch(child = vfork()) {
    case -1:
      strerr_die2sys(111,FATAL,"unable to fork: ");
    case 0:
      sig_pipedefault();
      execvp(*args,args);
      strerr_die4sys(111,FATAL,"unable to run ",arg,": ");
  }

  wait_pid(&wstat,child);
  if (wait_crashed(wstat))
    strerr_die4sys(111,FATAL,"child crashed in ",arg,": ");

  switch(wait_exitcode(wstat)) {
    case 64: case 65: case 70: case 76: case 77: case 78: case 112:
    case 100: _exit(100);
    case 0: break;
    default: _exit(111);
  }

  if (seek_begin(0) == -1)
    strerr_die2sys(111,FATAL,"unable to rewind input: ");
}

void dodata()
{
  int i;
  int j;
  i = 0;
  for (j = 0;j < data.len;++j)
    if (!data.s[j]) {
      if ((data.s[i] == '|') || (data.s[i] == '!'))
        doprogram(data.s + i);
      else if ((data.s[i] == '.') || (data.s[i] == '/')) {
        if (!stralloc_cats(&todo,data.s + i)) nomem();
        if (!stralloc_0(&todo)) nomem();
      }
      else if ((data.s[i] == '&') && (j - i < 900)) {
        if (!stralloc_cats(&todo,data.s + i)) nomem();
        if (!stralloc_0(&todo)) nomem();
      }
      i = j + 1;
    }
}

void dorecip(addr)
char *addr;
{

  if (!findtarget(0,"?",addr))
    if (gettarget(0,":",addr)) {
      dodata();
      return;
    }
  if (!stralloc_cats(&forward,addr)) nomem();
  if (!stralloc_0(&forward)) nomem();
}

void doorigrecip(addr)
char *addr;
{
  if (sender.len)
    if ((sender.len != 4) || byte_diff(sender.s,4,"#@[]"))
      if (gettarget(1,"?",addr))
        if (!stralloc_copy(&sender,&data)) nomem();
  if (!gettarget(1,":",addr))
    if (flagpassthrough)
      _exit(0);
    else
      strerr_die1x(100,"Sorry, no mailbox here by that name. (#5.1.1)");
  dodata();
}

stralloc recipient = {0};
int flagdefault = 0;

void main(argc,argv)
int argc;
char **argv;
{
  int opt;
  char *x;
  int i;

  sig_pipeignore();

  dtline = env_get("DTLINE");
  if (!dtline) dtline = "";

  x = env_get("SENDER");
  if (!x) x = "original envelope sender";
  if (!stralloc_copys(&sender,x)) nomem();

  if (!stralloc_copys(&forward,"")) nomem();
  if (!strset_init(&done)) nomem();

  while ((opt = getopt(argc,argv,"nNpPdD")) != opteof)
    switch(opt) {
      case 'n': flagdeliver = 0; break;
      case 'N': flagdeliver = 1; break;
      case 'p': flagpassthrough = 1; break;
      case 'P': flagpassthrough = 0; break;
      case 'd': flagdefault = 1; break;
      case 'D': flagdefault = 0; break;
      default: usage();
    }
  argv += optind;

  fncdb = *argv;
  if (!fncdb) usage();
  fdcdb = open_read(fncdb);
  if (fdcdb == -1) cdbreaderror();
  coe(fdcdb);

  if (flagdefault) {
    x = env_get("DEFAULT");
    if (!x) x = env_get("EXT");
    if (!x) strerr_die2x(100,FATAL,"$DEFAULT or $EXT must be set");
    if (!stralloc_copys(&recipient,x)) nomem();
    if (!stralloc_cats(&recipient,"@")) nomem();
    x = env_get("HOST");
    if (!x) strerr_die2x(100,FATAL,"$HOST must be set");
    if (!stralloc_cats(&recipient,x)) nomem();
    if (!stralloc_0(&recipient)) nomem();
    x = recipient.s;
  }
  else {
    x = env_get("RECIPIENT");
    if (!x) strerr_die2x(100,FATAL,"$RECIPIENT must be set");
  }
  if (!strset_add(&done,x)) nomem();
  doorigrecip(x);

  while (todo.len) {
    i = todo.len - 1;
    while ((i > 0) && todo.s[i - 1]) --i;
    todo.len = i;

    if (strset_in(&done,todo.s + i)) continue;

    x = alloc(str_len(todo.s + i) + 1);
    if (!x) nomem();
    str_copy(x,todo.s + i);
    if (!strset_add(&done,x)) nomem();

    x = todo.s + i;
    if (*x == 0)
      continue;
    else if ((*x == '.') || (*x == '/'))
      dofile(x);
    else
      dorecip(x + 1);
  }

  if (!forward.len) {
    if (!flagdeliver) {
      print("no forwarding\n");
      substdio_flush(subfderr);
    }
    _exit(flagpassthrough ? 99 : 0);
  }

  if (!stralloc_0(&sender)) nomem();

  if (!flagdeliver) {
    print("from <");
    printsafe(sender.s);
    print(">\n");
    while (forward.len) {
      i = forward.len - 1;
      while ((i > 0) && forward.s[i - 1]) --i;
      forward.len = i;
      print("to <");
      printsafe(forward.s + i);
      print(">\n");
    }
    substdio_flush(subfderr);
    _exit(flagpassthrough ? 99 : 0);
  }

  if (qmail_open(&qq) == -1)
    strerr_die2sys(111,FATAL,"unable to fork: ");
  qmail_puts(&qq,dtline);
  if (substdio_copy(&ssqq,&ssmess) != 0)
    strerr_die2sys(111,FATAL,"unable to read message: ");
  substdio_flush(&ssqq);
  qp[fmt_ulong(qp,qmail_qp(&qq))] = 0;

  qmail_from(&qq,sender.s);

  while (forward.len) {
    i = forward.len - 1;
    while ((i > 0) && forward.s[i - 1]) --i;
    forward.len = i;
    qmail_to(&qq,forward.s + i);
  }

  x = qmail_close(&qq);
  if (*x) strerr_die2x(*x == 'D' ? 100 : 111,FATAL,x + 1);
  strerr_die2x(flagpassthrough ? 99 : 0,"fastforward: qp ",qp);
}
